components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -154,6 +172,55 @@ open class GleanInternalAPI internal constructor () {
         return uploadEnabled
     }
 
+    /**
+     * Handles the changing of state when uploadEnabled changes.
+     *
+     * When disabling, all pending metrics, events and queued pings are cleared.
+     *
+     * When enabling, the core Glean metrics are recreated.
+     *
+     * If Glean is not initialized, this is a no-op.
+     */
+    private fun onChangeUploadEnabled(enabled: Boolean) {
+        if (enabled) {
+            initializeCoreMetrics(applicationContext!!)
+        } else {
+            clearMetrics()
+        }
+    }
+
+    /**
+     * Clear any pending metrics when telemetry is disabled.
+     */
+    @Suppress("EXPERIMENTAL_API_USAGE")
+    private fun clearMetrics() = Dispatchers.API.launch {
+        // There is only one metric that we want to survive after clearing all metrics:
+        // firstRunDate. Here, we store its value so we can restore it after clearing
+        // the metrics.
+        val firstRunDateMetric = GleanInternalMetrics.firstRunDate
```

Comment:
Note: this is not clearing sequence numbers, as we discussed yesterday

Reply:
It should be now.